lib/evaluate/fold.cc

```diff
@@ -156,35 +162,298 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
+    std::index_sequence<I...>) {
+  static_assert(
+      (... && IsSpecificIntrinsicType<TA>));  // TODO derived types for MERGE?
+  std::tuple<const std::optional<Scalar<TA>>...> scalars{
+      GetScalarConstantValue<TA>(*funcRef.arguments()[I]->value)...};
+  if ((... && std::get<I>(scalars).has_value())) {
+    if constexpr (std::is_same_v<WrapperType<TR, TA...>,
+                      ScalarFuncWithContext<TR, TA...>>) {
+      return Expr<TR>{Constant<TR>{func(context, *std::get<I>(scalars)...)}};
+    } else if constexpr (std::is_same_v<WrapperType<TR, TA...>,
+                             ScalarFunc<TR, TA...>>) {
+      return Expr<TR>{Constant<TR>{func(*std::get<I>(scalars)...)}};
+    }
+  }
+  // TODO: handle arrays when Constant<T> can represent them
+  return Expr<TR>{std::move(funcRef)};
+}
+
+template<typename TR, typename... TA>
+static Expr<TR> FoldElementalIntrinsic(FoldingContext &context,
+    FunctionRef<TR> &&funcRef, ScalarFunc<TR, TA...> func) {
+  return FoldElementalIntrinsicHelper<ScalarFunc, TR, TA...>(
+      context, std::move(funcRef), func, std::index_sequence_for<TA...>{});
+}
+template<typename TR, typename... TA>
+static Expr<TR> FoldElementalIntrinsic(FoldingContext &context,
+    FunctionRef<TR> &&funcRef, ScalarFuncWithContext<TR, TA...> func) {
+  return FoldElementalIntrinsicHelper<ScalarFuncWithContext, TR, TA...>(
+      context, std::move(funcRef), func, std::index_sequence_for<TA...>{});
+}
+}
+
+template<int KIND>
+Expr<Type<TypeCategory::Integer, KIND>> FoldOperation(FoldingContext &context,
+    FunctionRef<Type<TypeCategory::Integer, KIND>> &&funcRef) {
+  using namespace intrinsicHelper;
+  using T = Type<TypeCategory::Integer, KIND>;
+  for (std::optional<ActualArgument> &arg : funcRef.arguments()) {
     if (arg.has_value()) {
       *arg->value = FoldOperation(context, std::move(*arg->value));
     }
   }
   if (auto *intrinsic{std::get_if<SpecificIntrinsic>(&funcRef.proc().u)}) {
-    std::string name{intrinsic->name};
+    const std::string name{intrinsic->name};
     if (name == "kind") {
       if constexpr (common::HasMember<T, IntegerTypes>) {
-        return Expr<T>{args[0]->value->GetType()->kind};
+        return Expr<T>{funcRef.arguments()[0]->value->GetType()->kind};
       } else {
         common::die("kind() result not integral");
       }
     } else if (name == "len") {
       if constexpr (std::is_same_v<T, SubscriptInteger>) {
-        if (auto *charExpr{UnwrapExpr<Expr<SomeCharacter>>(*args[0]->value)}) {
+        if (auto *charExpr{UnwrapExpr<Expr<SomeCharacter>>(
+                *funcRef.arguments()[0]->value)}) {
           return std::visit([](auto &kx) { return kx.LEN(); }, charExpr->u);
         }
       } else {
         common::die("len() result not SubscriptInteger");
       }
-    } else {
-      // TODO: many more intrinsic functions
+    } else if (name == "iand") {
+      if (auto *x{std::get_if<BOZLiteralConstant>(
```

Comment:
First, you should loop here rather than copying and pasting code.  Second, since these conversions of typeless BOZ literals to the type of the function result are a common pattern, maybe the intrinsic function table lookup code in `intrinsics.cc` should be performing them.

Reply:
OK, I added a TODO for the second point and applied the first point for now.